File: app/src/main/java/com/cl/vlive/MainActivity.java
package com.cl.vlive;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.FileProvider;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.Date;

public class MainActivity extends AppCompatActivity {

    private static final int PICK_VIDEO_REQUEST = 1;
    private static final int REQUEST_PERMISSIONS = 2;
    
    private Button btnSelectVideo, btnConvert;
    private TextView tvStatus, tvStart, tvEnd;
    private SeekBar seekBarStart, seekBarEnd;
    
    private Uri videoUri;
    private long videoDuration = 0;
    private long startTime = 0;
    private long endTime = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnSelectVideo = findViewById(R.id.btnSelectVideo);
        btnConvert = findViewById(R.id.btnConvert);
        tvStatus = findViewById(R.id.tvStatus);
        tvStart = findViewById(R.id.tvStart);
        tvEnd = findViewById(R.id.tvEnd);
        seekBarStart = findViewById(R.id.seekBarStart);
        seekBarEnd = findViewById(R.id.seekBarEnd);

        btnSelectVideo.setOnClickListener(v -> selectVideo());
        btnConvert.setOnClickListener(v -> convertToLivePhoto());

        seekBarStart.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                startTime = (long) (progress / 100.0 * videoDuration);
                tvStart.setText("开始时间: " + formatTime(startTime));
                if (startTime >= endTime) {
                    endTime = startTime + 1000; // 确保结束时间大于开始时间
                    seekBarEnd.setProgress((int) (endTime * 100 / videoDuration));
                    tvEnd.setText("结束时间: " + formatTime(endTime));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {}

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        seekBarEnd.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                endTime = (long) (progress / 100.0 * videoDuration);
                tvEnd.setText("结束时间: " + formatTime(endTime));
                if (endTime <= startTime) {
                    startTime = endTime - 1000; // 确保开始时间小于结束时间
                    seekBarStart.setProgress((int) (startTime * 100 / videoDuration));
                    tvStart.setText("开始时间: " + formatTime(startTime));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {}

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        // 请求权限
        requestPermissions();
    }

    private void requestPermissions() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            if (checkSelfPermission(android.Manifest.permission.READ_EXTERNAL_STORAGE) != android.content.pm.PackageManager.PERMISSION_GRANTED ||
                    checkSelfPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE) != android.content.pm.PackageManager.PERMISSION_GRANTED) {
                requestPermissions(new String[]{android.Manifest.permission.READ_EXTERNAL_STORAGE, android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_PERMISSIONS);
            }
        }
    }

    private void selectVideo() {
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Video.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(intent, PICK_VIDEO_REQUEST);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PICK_VIDEO_REQUEST && resultCode == RESULT_OK && data != null) {
            videoUri = data.getData();
            tvStatus.setText("已选择视频: " + getVideoName(videoUri));
            
            // 获取视频时长
            videoDuration = getVideoDuration(videoUri);
            if (videoDuration > 0) {
                // 显示时间选择控件
                tvStart.setVisibility(View.VISIBLE);
                tvEnd.setVisibility(View.VISIBLE);
                seekBarStart.setVisibility(View.VISIBLE);
                seekBarEnd.setVisibility(View.VISIBLE);
                
                // 设置默认值
                startTime = 0;
                endTime = videoDuration;
                seekBarStart.setProgress(0);
                seekBarEnd.setProgress(100);
                tvStart.setText("开始时间: " + formatTime(startTime));
                tvEnd.setText("结束时间: " + formatTime(endTime));
            }
        }
    }

    private String getVideoName(Uri uri) {
        String[] projection = {MediaStore.Video.Media.DISPLAY_NAME};
        android.database.Cursor cursor = getContentResolver().query(uri, projection, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.Video.Media.DISPLAY_NAME);
            String name = cursor.getString(columnIndex);
            cursor.close();
            return name;
        }
        return "未知视频";
    }

    private long getVideoDuration(Uri uri) {
        String[] projection = {MediaStore.Video.Media.DURATION};
        android.database.Cursor cursor = getContentResolver().query(uri, projection, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.Video.Media.DURATION);
            long duration = cursor.getLong(columnIndex);
            cursor.close();
            return duration;
        }
        return 0;
    }

    private String formatTime(long milliseconds) {
        int seconds = (int) (milliseconds / 1000);
        int minutes = seconds / 60;
        seconds = seconds % 60;
        return String.format("%02d:%02d", minutes, seconds);
    }

    private void convertToLivePhoto() {
        if (videoUri == null) {
            Toast.makeText(this, "请先选择视频", Toast.LENGTH_SHORT).show();
            return;
        }

        new Thread(() -> {
            try {
                // 创建输出目录
                File outputDir = new File(getExternalFilesDir(Environment.DIRECTORY_PICTURES), "LivePhotos");
                if (!outputDir.exists()) {
                    outputDir.mkdirs();
                }

                // 生成文件名
                String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
                String fileName = "LIVE_" + timeStamp;

                // 提取视频片段
                File trimmedVideo = LivePhotoUtils.trimVideo(MainActivity.this, videoUri, outputDir, fileName, startTime, endTime);

                // 提取关键帧作为照片
                File photoFile = LivePhotoUtils.extractFrame(MainActivity.this, videoUri, outputDir, fileName);

                // 创建Live Photo包
                File livePhotoPackage = LivePhotoUtils.createLivePhotoPackage(outputDir, fileName, photoFile, trimmedVideo);

                runOnUiThread(() -> {
                    Toast.makeText(MainActivity.this, "Live Photo 已生成: " + livePhotoPackage.getAbsolutePath(), Toast.LENGTH_LONG).show();
                    tvStatus.setText("转换完成！");
                });

            } catch (Exception e) {
                Log.e("LivePic", "转换失败", e);
                runOnUiThread(() -> {
                    Toast.makeText(MainActivity.this, "转换失败: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                });
            }
        }).start();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_PERMISSIONS) {
            if (grantResults.length > 0 && grantResults[0] == android.content.pm.PackageManager.PERMISSION_GRANTED) {
                // 权限授予成功
            } else {
                Toast.makeText(this, "需要存储权限才能操作", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/cl/vlive/MainActivity.kt
package com.cl.vlive

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.widget.SeekBar
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.cl.vlive.databinding.ActivityMainBinding
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var selectedVideoUri: Uri? = null
    private var durationMs: Long = 0L
    private var lastExportUris: ArrayList<Uri> = arrayListOf()

    private val pickVideoLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
        if (uri != null) {
            selectedVideoUri = uri
            binding.videoView.setVideoURI(uri)
            binding.videoView.setOnPreparedListener { mp ->
                durationMs = mp.duration.toLong()
                updateRangeText()
                mp.isLooping = true
                binding.videoView.start()
            }
            binding.videoLabel.text = uri.toString()
            binding.exportButton.isEnabled = true
        }
    }

    private val permissionLauncher = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) {
        // no-op: continue with scoped storage even if partial denied
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        ensurePermissions()
        initRangeSeekBars()

        binding.pickButton.setOnClickListener {
            pickVideoLauncher.launch("video/*")
        }

        binding.exportButton.setOnClickListener {
            val uri = selectedVideoUri
            if (uri == null || durationMs <= 0) {
                toast("请先选择可用视频")
                return@setOnClickListener
            }
            exportLivePhoto(uri)
        }

        binding.shareButton.setOnClickListener {
            if (lastExportUris.isEmpty()) {
                toast("暂无可分享文件")
                return@setOnClickListener
            }
            shareLastExport()
        }

        updateRangeText()
    }

    private fun ensurePermissions() {
        val required = mutableListOf<String>()
        if (Build.VERSION.SDK_INT >= 33) {
            required += Manifest.permission.READ_MEDIA_VIDEO
            required += Manifest.permission.READ_MEDIA_IMAGES
        } else {
            required += Manifest.permission.READ_EXTERNAL_STORAGE
        }

        val missing = required.filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }
        if (missing.isNotEmpty()) {
            permissionLauncher.launch(missing.toTypedArray())
        }
    }

    private fun initRangeSeekBars() {
        binding.startSeek.progress = 0
        binding.endSeek.progress = 1000

        binding.startSeek.setOnSeekBarChangeListener(rangeListener)
        binding.endSeek.setOnSeekBarChangeListener(rangeListener)
    }

    private val rangeListener = object : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
            if (!fromUser) return

            var start = binding.startSeek.progress
            var end = binding.endSeek.progress

            if (start >= end) {
                if (seekBar?.id == binding.startSeek.id) {
                    start = (end - 10).coerceAtLeast(0)
                    binding.startSeek.progress = start
                } else {
                    end = (start + 10).coerceAtMost(1000)
                    binding.endSeek.progress = end
                }
            }

            if (durationMs > 0L) {
                val minGapProgress = ((1500f / durationMs) * 1000f).toInt().coerceAtLeast(1)
                val maxGapProgress = ((3000f / durationMs) * 1000f).toInt().coerceAtLeast(minGapProgress)
                val gap = end - start

                if (gap < minGapProgress) {
                    if (seekBar?.id == binding.startSeek.id) {
                        start = (end - minGapProgress).coerceAtLeast(0)
                        binding.startSeek.progress = start
                    } else {
                        end = (start + minGapProgress).coerceAtMost(1000)
                        binding.endSeek.progress = end
                    }
                } else if (gap > maxGapProgress) {
                    if (seekBar?.id == binding.startSeek.id) {
                        start = (end - maxGapProgress).coerceAtLeast(0)
                        binding.startSeek.progress = start
                    } else {
                        end = (start + maxGapProgress).coerceAtMost(1000)
                        binding.endSeek.progress = end
                    }
                }
            }

            updateRangeText()
            seekPreview()
        }

        override fun onStartTrackingTouch(seekBar: SeekBar?) = Unit
        override fun onStopTrackingTouch(seekBar: SeekBar?) = Unit
    }

    private fun seekPreview() {
        if (durationMs <= 0) return
        val startMs = progressToMs(binding.startSeek.progress)
        binding.videoView.seekTo(startMs.toInt())
    }

    private fun progressToMs(progress: Int): Long {
        return (durationMs * progress / 1000f).toLong()
    }

    private fun currentRange(): Pair<Long, Long> {
        val start = progressToMs(binding.startSeek.progress)
        val end = progressToMs(binding.endSeek.progress)
        return start to end
    }

    private fun updateRangeText() {
        val (start, end) = currentRange()
        binding.rangeText.text = "${formatMs(start)} - ${formatMs(end)}"
    }

    private fun formatMs(ms: Long): String {
        val total = ms / 1000
        val m = total / 60
        val s = total % 60
        val msPart = (ms % 1000) / 10
        return String.format("%02d:%02d.%02d", m, s, msPart)
    }

    private fun exportLivePhoto(videoUri: Uri) {
        val (startMs, endMs) = currentRange()
        binding.progressBar.visibility = android.view.View.VISIBLE
        binding.exportButton.isEnabled = false
        binding.shareButton.isEnabled = false

        thread {
            try {
                val exporter = LivePhotoExporter(this)
                val artifacts = exporter.export(videoUri, startMs, endMs)

                val imageUri = MediaStoreSaver.saveImage(this, artifacts.imageFile, artifacts.baseName)
                val videoUriOut = MediaStoreSaver.saveVideo(this, artifacts.videoFile, artifacts.baseName)
                lastExportUris = arrayListOf(imageUri, videoUriOut)

                runOnUiThread {
                    binding.progressBar.visibility = android.view.View.GONE
                    binding.exportButton.isEnabled = true
                    binding.shareButton.isEnabled = true
                    toast("导出成功：${artifacts.baseName}.JPG + ${artifacts.baseName}.MOV")
                }
            } catch (t: Throwable) {
                runOnUiThread {
                    binding.progressBar.visibility = android.view.View.GONE
                    binding.exportButton.isEnabled = true
                    toast("导出失败: ${t.message}")
                }
            }
        }
    }

    private fun shareLastExport() {
        val intent = Intent(Intent.ACTION_SEND_MULTIPLE).apply {
            type = "*/*"
            putParcelableArrayListExtra(Intent.EXTRA_STREAM, lastExportUris)
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
        startActivity(Intent.createChooser(intent, "分享 Live Photo 文件"))
    }

    private fun toast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_LONG).show()
    }
}

